pkg/client/fwd_pipe.go

```diff
@@ -29,18 +39,67 @@ func (c *Client) SetFwdPipeFromBytes(binBytes, p4infoBytes []byte) error {
 	_, err := c.SetForwardingPipelineConfig(context.Background(), req)
 	if err == nil {
 		c.p4Info = p4Info
+		return &FwdPipeConfig{
+			P4Info:         p4Info,
+			P4DeviceConfig: binBytes,
+			Cookie:         cookie,
+		}, nil
 	}
-	return err
+
+	return nil, err
 }
 
-func (c *Client) SetFwdPipe(binPath string, p4infoPath string) error {
+func (c *Client) SetFwdPipe(binPath string, p4infoPath string, cookie uint64) (*FwdPipeConfig, error) {
 	binBytes, err := ioutil.ReadFile(binPath)
 	if err != nil {
-		return fmt.Errorf("error when reading binary device config: %v", err)
+		return nil, fmt.Errorf("error when reading binary device config: %v", err)
 	}
 	p4infoBytes, err := ioutil.ReadFile(p4infoPath)
 	if err != nil {
-		return fmt.Errorf("error when reading P4Info text file: %v", err)
+		return nil, fmt.Errorf("error when reading P4Info text file: %v", err)
+	}
+	return c.SetFwdPipeFromBytes(binBytes, p4infoBytes, cookie)
+}
+
+// GetFwdPipe retrieves the current pipeline config used in the remote switch.
+//
+// responseType is oneof:
+//  "ALL", "COOKIE_ONLY", "P4INFO_AND_COOKIE", "DEVICE_CONFIG_AND_COOKIE"
+// See https://p4.org/p4runtime/spec/v1.3.0/P4Runtime-Spec.html#sec-getforwardingpipelineconfig-rpc
+func (c *Client) GetFwdPipe(responseType string) (*FwdPipeConfig, error) {
```

Comment:
I think we could introduce a new type for the parameter, and remove the error checking below:

```go
type GetFwdPipeResponseType string
const (
  GetFwdPipeAll GetFwdPipeResponseType = "ALL"
  GetFwdPipeCookieOnly = "COOKIE_ONLY"
  GetFwdPipeP4InfoAndCookie = "P4INFO_AND_COOKIE"
  GetFwdPipeDeviceConfigAndCookie = "DEVICE_CONFIG_AND_COOKIE"
)
```
These constants could even be (shorter) aliases for `GetForwardingPipelineConfigRequest_ResponseType` constants generated by protoc:

```go
const (
  GetFwdPipeAll = p4_v1.GetForwardingPipelineConfigRequest_ALL
 //  ...
)
```

Or, the user could use the protoc constants directly, since they are not that long, but maybe that's a bit less practical, since they are not defined in this package.

Reply:
Good idea! Will change!